File: Pawsome-Windows/Pawsome.Core/Models/AppUser.cs
using Pawsome.Core.Firestore;

namespace Pawsome.Core.Models;

/// <summary>A signed-in Pawsome user (the `users/{uid}` document).</summary>
public sealed class AppUser
{
    public required string Uid { get; init; }
    public string Username { get; set; } = "User";
    public string? ProfilePic { get; set; }
    public int UserNumber { get; init; }

    public static AppUser FromFirestore(string uid, IReadOnlyDictionary<string, object?> data) => new()
    {
        Uid = uid,
        Username = data.GetString("username") ?? "User",
        ProfilePic = data.GetString("profilePic"),
        UserNumber = (int)data.GetLong("userNumber"),
    };
}

File: Android/app/src/main/java/com/pawsome/app/model/AppUser.kt
package com.example.pawsome.model

import com.google.firebase.firestore.DocumentSnapshot

data class AppUser(
    val uid: String,
    var username: String,
    var profilePic: String?,
    val userNumber: Int,
    val loginMethod: String,
    val joinedOnMillis: Long,
) {
    companion object {
        fun fromDocument(document: DocumentSnapshot) = AppUser(
            uid = document.id,
            // Read both spellings so existing profiles created by older builds
            // continue to work after the field is corrected to "Username".
            username = document.getString("Username")
                ?: document.getString("Usename")
                ?: "User",
            profilePic = document.getString("ProfilePic"),
            userNumber = (document.getLong("UserID") ?: 0L).toInt(),
            loginMethod = document.getString("LoginMethod") ?: "Unknown",
            joinedOnMillis = document.getTimestamp("JoinedOn")?.toDate()?.time ?: 0L,
        )
    }
}
